rate_limit.py

```diff
@@ -0,0 +1,103 @@
+"""
+This module contains functions for rate limiting requests.
```

Comment:
please explain here why we needed this custom rate limiter when there are opensource libraries already
https://pypi.org/project/ratelimiter/

Reply:
I'll update to this
```
This module contains functions for rate limiting requests.

The rate limiting system operates on two levels:
1. User-level rate limiting: Each user (identified by a token) has a
   configurable minimum interval between requests.

2. System-wide rate limiting: There is a global limit on the total number of 
   requests across all users within a specified time period.
```